Allow authenticating to pull images everywhere it is missing and release fix 5.21.3

sampaccoud opened the pull request:
## Purpose

Dockerhub recently enforced rate limits for anonymous users which sometimes block us from downloading Docker images. We must allow authenticating on all image pulls. It was done partially but some images are still pulled anonymously and we get blocked very often in our deployments.

## Purpose

Add credentials to objects definitions everywhere an image is pulled from an external registry and it is missing.

:warning: the nextcloud image is pulled from OpenShift's local registry and is therefore not treated here. It must be refactored separately.

sampaccoud committed:
✨(apps) allow authenticating to pull images everywhere it is missing

Dockerhub recently enforced rate limits for anonymous users which
sometimes block us from downloading Docker images. We must allow
authenticating on all image pulls.

lunika commented:
Why not also adding the secret template ? It should be created only when needed.

jmaupetit commented on apps/ashley/templates/services/nginx/dc.yml.j2:
```diff
@@ -32,6 +32,10 @@ spec:
               topologyKey: kubernetes.io/hostname
+{% if ashley_nginx_image_pull_secret_name is defined and ashley_nginx_image_pull_secret_name is not none %}
+      imagePullSecrets:
+        - name: "{{ ashley_nginx_image_pull_secret_name }}"
```
Shouldn't we fallback to a global secret (e.g. organization's dockerhub account)?

```suggestion
        - name: "{{ ashley_nginx_image_pull_secret_name | default(default_image_pull_secret_name) }}"
```

sampaccoud replied:
good idea but the `if` has to be modified as well :sweat_smile:

sampaccoud replied:
@jmaupetit I refactored to allow a `default_image_pull_secret_name` variable as you suggested :pray:

madmatah commented on apps/learninglocker/templates/services/xapi/dc.yml.j2:
```diff
@@ -35,6 +35,10 @@ spec:
                         - "learninglocker-xapi-{{ deployment_stamp }}"
                 topologyKey: kubernetes.io/hostname
       containers:
+{% if learninglocker_xapi_image_pull_secret_name is defined and learninglocker_xapi_image_pull_secret_name is not none %}
```
:warning: This `imagePullsecrets` block has been mixed with the `containers` block.

sampaccoud replied:
good catch. I corrected!

madmatah commented on apps/marsha/templates/services/app/cronjob_check_live_state.yml.j2:
```diff
@@ -25,6 +25,10 @@ spec:
             version: "{{ marsha_image_tag }}"
             deployment_stamp: "{{ deployment_stamp }}"
         spec:
+{% if marsha_image_pull_secret_name is defined and marsha_image_pull_secret_name is not none %}
```
Missing indentation ? It should be on the same level than the containers definition.

sampaccoud committed:
🔖(patch) bump release to 5.21.3

Fixed

- Fix Docker lint error by passing --no-cache-dir option on pip install
- Allow authenticating to pull images everywhere it is missing

Removed

- Forum app build config and directly use image from external registry

jmaupetit commented on apps/ashley/templates/services/app/dc.yml.j2:
```diff
@@ -32,9 +32,10 @@ spec:
                 topologyKey: kubernetes.io/hostname
-{% if ashley_image_pull_secret_name is defined and ashley_image_pull_secret_name is not none %}
+{% set image_pull_secret_name = ashley_image_pull_secret_name or default_image_pull_secret_name %}
+{% if image_pull_secret_name is defined and image_pull_secret_name is not none %}
```
If it is set, it will always be defined right?

jmaupetit replied:
```suggestion
{% if image_pull_secret_name is not none %}
```

sampaccoud replied:
true... :pray:

jmaupetit commented on apps/ashley/templates/services/app/job_01_collectstatic.yml.j2:
```diff
@@ -22,9 +22,10 @@ spec:
     spec:
-{% if ashley_image_pull_secret_name is defined and ashley_image_pull_secret_name is not none %}
+{% set image_pull_secret_name = ashley_image_pull_secret_name or default_image_pull_secret_name %}
++{% if image_pull_secret_name is defined and image_pull_secret_name is not none %}
```
```suggestion
{% if image_pull_secret_name is defined and image_pull_secret_name is not none %}
```

jmaupetit replied:
This error is repeated multiple times.

sampaccoud replied:
yes it was a regex search/replace :pray:

jmaupetit commented on apps/etherpad/templates/services/postgresql/dc.yml.j2:
```diff
@@ -20,6 +20,10 @@ spec:
+{% if etherpad_database_pull_secret_name is defined and etherpad_database_pull_secret_name is not none %}
+      imagePullSecrets:
+        - name: "{{ etherpad_database_pull_secret_name }}"
+{% endif %}
```
```suggestion
{% set image_pull_secret_name = etherpad_database_pull_secret_name or default_image_pull_secret_name %}
{% if image_pull_secret_name is not none %}
      imagePullSecrets:
        - name: "{{ image_pull_secret_name }}"
{% endif %}
```

jmaupetit commented on apps/hello/templates/services/app/cronjob_hello.yml.j2:
```diff
@@ -23,6 +23,10 @@ spec:
+{% if busybox_image_pull_secret_name is defined and busybox_image_pull_secret_name is not none %}
+          imagePullSecrets:
+            - name: "{{ busybox_image_pull_secret_name }}"
+{% endif %}
```
Ditto

jmaupetit commented on apps/learninglocker/templates/services/app/job_storage.yml.j2:
```diff
@@ -20,6 +20,11 @@ spec:
         job_stamp: "{{ job_stamp }}"
         deployment_stamp: "{{ deployment_stamp }}"
     spec:
+{% set image_pull_secret_name = busybox_image_pull_secret_name or default_image_pull_secret_name %}
```
```suggestion
{% set image_pull_secret_name = learninglocker_image_pull_secret_name or default_image_pull_secret_name %}
```

sampaccoud replied:
This job uses the busybox image so I think this line was correct.

sampaccoud committed:
⚡️(Dockerfile) pass --no-cache-dir option on pip install

This improvement was requested by the latest lint Docker image.
It should optimise the image size by disabling pip cache.

sampaccoud committed:
🔥(forum) remove build config and point to external image directly

The build config was added in case we wanted to install additional
packages to the live image. This is not used and we will soon get
rid of the OpenShift registry to only use images from an external
registry.

jmaupetit reviewed (approved):
![](https://media.giphy.com/media/26gsjCZpPolPr3sBy/giphy-downsized.gif)


:warning: :warning: Your release commit does not reflect all changes from this release. :warning: :warning:

sampaccoud commented:
@jmaupetit :pray:
Just :wrench: it!
